onceop annotation and linter

eed3si9n opened the pull request:
Ref https://github.com/scala/bug/issues/12803

**Problem**
There are certain classes, such as `Iterator` where the number of times you can call a method is restricted to once. This kind of restriction is currently not expressed as type, and it can catch users off guard.

**Solution**
This introduces a new annotation called `@onceop(...)`, which can mark methods that can be called once per instance per onceop group. For example, all `Iterator` methods except for `hasNext` and `next` are marked with `@onceop("all")` such that calling `take(2)`, and then calling any other method would now result to a warning.

Since I am tracking this within a scope, this is more as a lightweight check, and by no means exhaustive.
It doesn't check for example passing an iterator to a function, which inside could make a onceop call or not.

eed3si9n committed:
onceop annotation and linter

**Problem**
There are certain classes, such as `Iterator` where
the number of times you can call a method is restricted to once.
This kind of restriction is currently not expressed as type,
and it can catch users off guard.

**Solution**
This introduces a new annotation called `@onceop(...)`,
which can mark methods that can be called once per instance
per onceop group. For example, all `Iterator` methods
except for `hasNext` and `next` are marked with `@onceop("default")`
such that calling `take(2)` twice would now result to a warning.

Since I am tracking this within a scope, this is more as a
lightweight check, and by no means exhaustive.
It doesn't check for example passing an iterator to a function,
which inside could make a onceop call or not.

som-snytt commented:
Isn't Onceop the name of an ancient Egyptian pharaoh?

eed3si9n commented:
I have another impl that puts the annotation at the type instead - https://github.com/scala/scala/pull/10428

eed3si9n commented:
This found a bug in unit test by the way.
https://github.com/scala/scala/blob/6243e902928c344fb0e82e21120bb257f08a2af2/test/junit/scala/collection/IteratorTest.scala#L277-L281

It's not safe to call `contains(...)` method more than once.

eed3si9n committed:
Fix the test to not specify undefined behavior

som-snytt commented:
@eed3si9n very satisfying catch!

som-snytt commented:
I put the test tweak on a separate PR. The use of `isEmpty` on an invalid iterator must also be wrong, even though `isEmpty` is just `hasNext`; and same for `hasNext` and `next`; the wording in the doc is slightly ambiguous.

eed3si9n commented:
> But the reason the test succeeds is that contains is using `IterableOnce.exists`, which advances the iterator (itself) to look for the element, so it finds 5 then 6 but then is past 4 already. I wonder whether that is a legitimate test if the coder says it is. (That is, if tests are documentation.) 🤔

I think the multiple `contains` started showing up in https://github.com/scala/scala/pull/4075. From the fact that it is testing 6 then 4, I think it's clear that the author wrote it intentionally as a documentation of the internal behavior. In the 2023 hindsight I am claiming that the test should be specific enough to cover the intended behavior, but not document the undefined behavior.

> The use of `isEmpty` on an invalid iterator must also be wrong, even though `isEmpty` is just `hasNext`; and same for `hasNext` and `next`; the wording in the doc is slightly ambiguous.

Yea, you're right on both points.

1. I was being lenient with `isEmpty` but it should either be documented ok, or be marked once-op.
2. Once a symbol is once-op'ed, I guess any operation including `hasNext` should be considered unsafe.

eed3si9n committed:
all group dicards for all

eed3si9n commented:
With new stricter check, we find more bugs:

```scala
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/library/scala/collection/SortedMap.scala:125:13: call to an affine reference value i is unsafe after a once-op has been called
[error]       if (i.isEmpty) coll
[error]             ^
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/library/scala/collection/SortedMap.scala:126:25: call to an affine reference value i is unsafe after a once-op has been called
[error]       else rangeUntil(i.next())
[error]                         ^
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/library/scala/collection/SortedSet.scala:107:13: call to an affine reference value i is unsafe after a once-op has been called
[error]       if (i.isEmpty) coll
[error]             ^
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/library/scala/collection/SortedSet.scala:108:25: call to an affine reference value i is unsafe after a once-op has been called
[error]       else rangeUntil(i.next())
[error]                         ^
```

Hopefully I can fix them by rewriting to use `!hasNext` instead.

eed3si9n committed:
Fix isEmpty usages

som-snytt commented:
The multiple contains are from the original test (obvs I checked), but yes it looks like I added an extra check, probably because I was asking, What is this intended to verify?
```
  def check_range3: Int = {
    def trues(xs: List[Boolean]) = xs.foldLeft(0)((a, b) => if (b) a+1 else a)
    val r1 = Iterator.range(0, 10)
    val xs1 = List(r1 contains 5, r1 contains 6)
    val r2a = Iterator.range(0, 10, 2)
    val r2b = Iterator.range(0, 10, 2)
    val xs2 = List(r2a contains 5, r2b contains 6)
    val r3 = Iterator.range(0, 10, 11)
    val xs3 = List(r3 contains 5, r3 contains 6)
    // 2 + 1 + 0
    trues(xs1) + trues(xs2) + trues(xs3)
  }
```
Interestingly, in 2007, `scala.Range` was added to encapsulate `Iterator.range`! Back then, it seems `range` was more wide-ranging.

Oh, I see it originally defined
```
  def contains(x: Int): Boolean =
    Iterator.range(0, length) exists { x == start + _ * step }
```
that explains everything.

som-snytt commented:
Without trying to understand `rangeTo`, calling `isEmpty` is OK, it's just `hasNext`, but you can't use it after other methods (one might call them "terminal" methods in the lingo of Java streams).

eed3si9n committed:
More test fixes

eed3si9n commented:
> but you can't use it after other methods (one might call them "terminal" methods in the lingo of Java streams).

https://github.com/scala/scala/blob/6243e902928c344fb0e82e21120bb257f08a2af2/test/junit/scala/collection/immutable/RangeTest.scala#L56-L63

There was a bug of that variant where `next()` is called on a spent iterator. Since drop returns an iterator I fixed it by calling `next()` on it:

```scala
    val dropped = it.drop(1)

    assertEquals(11, dropped.next())
```

som-snytt commented:
I prefer your term "spent" iterator to Java's "consumed" stream.
